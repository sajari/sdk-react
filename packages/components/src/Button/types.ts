import { AriaButtonProps } from '@react-types/button';
import { PropsWithAs } from '@sajari/react-sdk-utils';

import { BoxProps } from '../Box';

export type ButtonAppearance = 'default' | 'primary' | 'link' | 'subtle-link' | 'none';
export type ButtonSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl';
export type ButtonSpacing = 'compact' | 'default' | 'none';

interface Props extends BoxProps {
  /** The base styling to apply to the button */
  appearance?: ButtonAppearance;
  /** Set how much spacing the button should have */
  spacing?: ButtonSpacing;
  /** The size to apply to the button */
  size?: ButtonSize;
  /** Provides a url for buttons being used as a link */
  loading?: boolean;
  /** Set if the button should be full width */
  fullWidth?: boolean;
  /** The classname for pressed button */
  pressedClassName?: string;
}

export interface ButtonProps extends PropsWithAs<Props>, AriaButtonProps {}
