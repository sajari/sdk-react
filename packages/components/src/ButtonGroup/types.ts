import { PropsWithAs } from '@sajari/react-sdk-utils';
import React from 'react';

import { BoxProps } from '../Box';

export interface Props extends BoxProps {
  /** Should the children be attached to each other */
  attached?: boolean;
  /** Make children fullWidth */
  fullWidth?: boolean;
  /** Display children horizontally */
  inline?: boolean;
  /** Children */
  children: React.ReactNode;
}

export type UseButtonGroupStylesParams = Pick<Props, 'attached' | 'inline'>;

export type ButtonGroupProps = PropsWithAs<Props>;
