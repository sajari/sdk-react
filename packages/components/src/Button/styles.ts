import { css } from '@emotion/core';
import { useTheme } from 'sajari-react-sdk-styles';
import tw, { TwStyle } from 'twin.macro';

import { ButtonProps } from './types';

const useButtonStyles = ({
  size,
  fullWidth,
  appearance,
  spacing,
  disabled,
  loading,
  pressed,
  focused,
  hovered,
}: ButtonProps & { pressed: boolean; hovered: boolean; focused: boolean }) => {
  const theme = useTheme();
  const isLink = (['link', 'subtle-link'] as Array<ButtonProps['apperance']>).includes(appearance);
  const styles: (string | TwStyle)[] = [];

  styles.push(
    tw`items-center justify-center no-underline transition duration-150 ease-in-out border border-transparent border-solid cursor-pointer focus:outline-none`,
  );

  if (!isLink) {
    styles.push(tw`font-medium rounded-md select-none`);
  }

  styles.push(fullWidth ? tw`flex` : tw`inline-flex`);

  switch (size) {
    case 'xl':
      if (!isLink) {
        styles.push(tw`text-xl`);
      }
      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-8`);
        } else {
          styles.push(tw`px-12 py-4`);
        }
      }
      break;

    case 'lg':
      if (!isLink) {
        styles.push(tw`text-lg`);
      }
      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-6`);
        } else {
          styles.push(tw`px-8 py-3`);
        }
      }
      break;

    case 'sm':
      styles.push(tw`text-sm`);

      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-2`);
        } else {
          styles.push(tw`px-3 py-1`);
        }
      }
      break;

    case 'xs':
      styles.push(tw`text-xs`);

      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-2`);
        } else {
          styles.push(tw`px-3 py-1`);
        }
      }
      break;

    default:
    case 'md':
      if (!isLink) {
        styles.push(tw`text-base`);
      }
      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-3`);
        } else {
          styles.push(tw`px-4 py-2`);
        }
      }
      break;
  }

  if (disabled) {
    styles.push(tw`text-gray-400 cursor-not-allowed`);

    switch (appearance) {
      case 'default':
        styles.push(tw`bg-gray-100 border-gray-200`);
        break;

      default:
        break;
    }

    return css(styles);
  }

  switch (appearance) {
    // TODO: test theming for primary appearance, we might want to change them later
    case 'primary':
      styles.push(`color: ${theme.color.primary.text};`);

      if (focused || pressed) {
        styles.push(`background: ${theme.color.primary.active};`);
      } else if (hovered) {
        styles.push(`background: ${theme.color.primary.active};`);
      } else {
        styles.push(`background: ${theme.color.primary.base};`);
      }
      break;

    case 'default':
      styles.push(
        tw`text-gray-500 bg-white border-gray-200 shadow-sm focus:border-gray-400 hover:bg-gray-100 focus:bg-white focus:shadow-outline-gray`,
      );
      if (pressed) {
        styles.push(tw`bg-gray-100`);
      }
      break;

    case 'link':
      styles.push(
        tw`no-underline bg-transparent hover:underline focus:underline focus:shadow-outline focus:shadow-outline-blue`,
      );
      styles.push({ color: pressed ? theme.color.primary.active : theme.color.primary.base });

      if (!loading && !pressed) {
        styles.push(`&:hover, &:focus { color: ${theme.color.primary.active} }`);
      }
      break;

    case 'subtle-link':
      styles.push(
        tw`no-underline bg-transparent hover:underline focus:underline focus:shadow-outline focus:shadow-outline-gray`,
      );

      styles.push(pressed ? tw`text-gray-700` : tw`text-gray-500`);

      if (!loading && !pressed) {
        styles.push(tw`hover:text-gray-700 focus:text-gray-700`);
      }
      break;
    default:
      break;
  }

  return css(styles);
};

export default useButtonStyles;
