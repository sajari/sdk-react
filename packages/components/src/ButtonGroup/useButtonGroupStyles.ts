/* eslint-disable no-nested-ternary */
import { css } from '@emotion/core';
import tw from 'twin.macro';
import { UseButtonGroupStylesParams } from './types';

export function useButtonGroupStyles({ attached, inline }: UseButtonGroupStylesParams) {
  const styles = [tw`inline-flex`];
  styles.push(inline ? tw`space-x-4 flex-row` : tw`space-y-2 flex-col`);
  styles.push(attached ? (inline ? tw`-space-x-px` : tw`-space-y-px`) : tw``);
  return css(styles);
}
