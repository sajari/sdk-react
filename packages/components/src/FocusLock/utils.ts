const focusableElements = [
  'a[href]',
  'area[href]',
  'button:not([disabled])',
  'embed',
  'iframe',
  'input:not([disabled])',
  'object',
  'select:not([disabled])',
  'textarea:not([disabled])',
  '*[tabindex]:not([aria-disabled])',
  '*[contenteditable]',
];

const focusableElSelector = focusableElements.join();

export function getFocusables(element: HTMLElement | null, keyboardOnly = false) {
  if (element === null) {
    return [];
  }

  let elements = Array.from(element.querySelectorAll(focusableElSelector));

  // Filter out elements with display: none
  elements = elements.filter((focusableEl) => window.getComputedStyle(focusableEl).display !== 'none');

  if (keyboardOnly === true) {
    elements = elements.filter((focusableEl) => focusableEl.getAttribute('tabindex') !== '-1');
  }

  return elements;
}
