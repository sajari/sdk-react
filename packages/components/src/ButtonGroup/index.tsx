/* eslint-disable no-nested-ternary */
/** @jsx jsx */
import { jsx } from '@emotion/core';
import { __DEV__, cleanChildren, getStylesObject } from '@sajari/react-sdk-utils';
import React, { cloneElement } from 'react';
import tw, { styled } from 'twin.macro';

import Box from '../Box';
import { useButtonGroupStyles } from './styles';
import { ButtonGroupProps } from './types';

// TODO: remove "any"
const StyledBox = styled<any>(Box, { shouldForwardProp: (prop) => prop !== 'attached' && prop !== 'inline' })<
  Pick<ButtonGroupProps, 'attached' | 'inline'>
>`
  & > button {
    :first-of-type {
      ${(props) =>
        props.attached
          ? props.inline
            ? tw`rounded-none rounded-l-md after:(rounded-none rounded-l-lg)`
            : tw`rounded-none rounded-t-md after:(rounded-none rounded-t-lg)`
          : tw``}
    }

    :last-of-type {
      ${(props) =>
        props.attached
          ? props.inline
            ? tw`rounded-none rounded-r-md after:(rounded-none rounded-r-lg)`
            : tw`rounded-none rounded-b-md after:(rounded-none rounded-b-lg)`
          : tw``}
    }

    :not(:last-of-type):not(:first-of-type) {
      ${(props) => (props.attached ? tw`rounded-none  after:(rounded-none)` : tw``)};
    }

    ${tw`focus:z-10`};
  }
`;

const ButtonGroup = React.forwardRef((props: ButtonGroupProps, ref?: React.Ref<HTMLDivElement>) => {
  const {
    attached = false,
    fullWidth,
    inline = true,
    children,
    as = 'div',
    disableDefaultStyles = false,
    styles: stylesProp,
    ...rest
  } = props;
  const styles = getStylesObject({ container: useButtonGroupStyles({ attached, inline }) }, disableDefaultStyles);
  const validChildren = cleanChildren(children);

  // TODO: handle case where child is Tooltip
  const clones = validChildren.map((child) =>
    cloneElement(child, {
      fullWidth,
      disableDefaultStyles,
      ...child.props,
    }),
  );

  if (disableDefaultStyles) {
    return (
      <Box as={as} ref={ref} css={stylesProp} {...rest}>
        {clones}
      </Box>
    );
  }

  return (
    <StyledBox attached={attached} inline={inline} as={as} ref={ref} css={[styles.container, stylesProp]} {...rest}>
      {clones}
    </StyledBox>
  );
});

if (__DEV__) {
  ButtonGroup.displayName = 'ButtonGroup';
}

export default ButtonGroup;
export type { ButtonGroupProps };
