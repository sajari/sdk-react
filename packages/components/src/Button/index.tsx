/** @jsx jsx */
import { jsx } from '@emotion/core';
import { useButton } from '@react-aria/button';
import { useFocus, useHover } from '@react-aria/interactions';
import { mergeProps } from '@react-aria/utils';
import { getStylesObject, mergeRefs } from '@sajari/react-sdk-utils';
import classnames from 'classnames';
import React, { useRef } from 'react';

import Box from '../Box';
import useButtonStyles from './styles';
import { ButtonProps } from './types';

const Button = React.forwardRef((props: ButtonProps, ref?: React.Ref<HTMLElement>) => {
  const {
    // TODO: handle the state later as we might need a Spinner/Icon component
    loading = false,
    as = 'button',
    appearance = 'default',
    size = 'md',
    spacing = 'default',
    disabled,
    fullWidth = false,
    onClick,
    children,
    className,
    styles: stylesProps,
    pressedClassName = '',
    disableDefaultStyles = false,
    ...rest
  } = props;

  const [focused, setFocused] = React.useState(false);
  const buttonRef = useRef<HTMLElement | null>(null);
  const ownRef = mergeRefs(buttonRef, ref);

  const { buttonProps, isPressed: pressed } = useButton(
    { ...rest, isDisabled: disabled, elementType: as, onPress: onClick },
    buttonRef,
  );
  const { hoverProps, isHovered: hovered } = useHover({
    isDisabled: disabled,
  });
  const { focusProps } = useFocus({
    isDisabled: disabled,
    onFocus: () => setFocused(true),
    onBlur: () => setFocused(false),
  });

  const { styles: containerStyles, focusRingProps } = useButtonStyles({
    pressed,
    appearance,
    disabled,
    fullWidth,
    loading,
    size,
    spacing,
    focused,
    hovered,
  });

  const customProps = mergeProps(buttonProps, focusProps, hoverProps, focusRingProps);
  const styles = getStylesObject({ container: containerStyles }, disableDefaultStyles);

  return (
    <Box
      as={as}
      ref={ownRef}
      {...customProps}
      css={[styles.container, stylesProps]}
      className={classnames(className, {
        [pressedClassName]: pressed,
      })}
    >
      {children}
    </Box>
  );
});

export default Button;
export type { ButtonProps };
