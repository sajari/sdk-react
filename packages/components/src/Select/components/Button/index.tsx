import { mergeProps } from '@react-aria/utils';
import { getStylesObject } from '@sajari/react-sdk-utils';
import * as React from 'react';

import Box from '../../../Box';
import { useSelectContext } from '../../context';
import { useButtonStyles } from './styles';

const Button = () => {
  const {
    disabled,
    getToggleButtonProps,
    id,
    invalid,
    items,
    selectedItems,
    autofocus,
    size,
    text,
    disableDefaultStyles,
    customClassNames,
  } = useSelectContext();
  const { styles: buttonStyles, focusRingProps, focusRingStyles } = useButtonStyles({ size, disabled, invalid });
  const styles = getStylesObject(buttonStyles, disableDefaultStyles);
  const ref = React.useRef<HTMLButtonElement>(null);

  React.useEffect(() => {
    if (autofocus && ref?.current) {
      ref.current.focus();
    }
  }, []);

  const children: React.ReactNode = React.useMemo(() => {
    if (typeof text === 'function') {
      const selected = items.filter(([v]) => selectedItems.includes(v.toString())).map(([, label]) => label);
      return text(selected);
    }

    if (typeof text !== 'undefined') {
      return text;
    }

    if (selectedItems.length === 1) {
      const [selectedItem] = selectedItems;
      const item = items.find(([v]) => selectedItem === v.toString());

      if (item) {
        const [, t] = item;
        return t;
      }
    }

    if (selectedItems.length > 1) {
      return `${selectedItems.length} selected`;
    }

    return 'Select an option';
  }, [JSON.stringify(selectedItems), text]);

  return (
    <Box
      as="button"
      type="button"
      css={[styles.container, focusRingStyles]}
      className={customClassNames.buttonClassName}
      {...mergeProps(focusRingProps, getToggleButtonProps({ disabled, ref }), { id })}
    >
      {children}
    </Box>
  );
};

export default Button;
