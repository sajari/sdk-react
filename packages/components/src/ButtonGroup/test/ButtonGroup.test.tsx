import React from 'react';

import Button from '../../Button';
import { render } from '../../test/utils';
import ButtonGroup from '..';

describe('ButtonGroup', () => {
  it('attached and inline', () => {
    const { asFragment } = render(
      <ButtonGroup attached>
        <Button>Button</Button>
        <Button>Button</Button>
        <Button>Button</Button>
      </ButtonGroup>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('attached and not inline', () => {
    const { asFragment } = render(
      <ButtonGroup attached>
        <Button>Button</Button>
        <Button>Button</Button>
        <Button>Button</Button>
      </ButtonGroup>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('not attached and inline', () => {
    const { asFragment } = render(
      <ButtonGroup>
        <Button>Button</Button>
        <Button>Button</Button>
      </ButtonGroup>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('not attached and not inline', () => {
    const { asFragment } = render(
      <ButtonGroup inline={false}>
        <Button>Button</Button>
        <Button>Button</Button>
      </ButtonGroup>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
