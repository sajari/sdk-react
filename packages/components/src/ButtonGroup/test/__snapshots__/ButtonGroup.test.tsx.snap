// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup attached and inline 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-0>button:first-of-type {
  border-radius: 0;
  border-top-left-radius: 0.375rem;
  border-bottom-left-radius: 0.375rem;
}

.emotion-0>button:first-of-type:after {
  border-radius: 0;
  border-top-left-radius: 0.5rem;
  border-bottom-left-radius: 0.5rem;
}

.emotion-0>button:last-of-type {
  border-radius: 0;
  border-top-right-radius: 0.375rem;
  border-bottom-right-radius: 0.375rem;
}

.emotion-0>button:last-of-type:after {
  border-radius: 0;
  border-top-right-radius: 0.5rem;
  border-bottom-right-radius: 0.5rem;
}

.emotion-0>button:not(:last-of-type):not(:first-of-type) {
  border-radius: 0;
}

.emotion-0>button:not(:last-of-type):not(:first-of-type):after {
  border-radius: 0;
}

.emotion-0>button:focus {
  z-index: 10;
}

.emotion-0>:not(template)~:not(template) {
  --space-x-reverse: 0;
  margin-right: calc(1rem * var(--space-x-reverse));
  margin-left: calc(1rem * calc(1 - var(--space-x-reverse)));
}

.emotion-0>:not(template)~:not(template) {
  --space-x-reverse: 0;
  margin-right: calc(-1px * var(--space-x-reverse));
  margin-left: calc(-1px * calc(1 - var(--space-x-reverse)));
}

.emotion-2 {
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1.5;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,-webkit-transform;
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  border-width: 1px;
  border-color: transparent;
  border-style: solid;
  cursor: pointer;
  font-family: inherit;
  font-weight: 500;
  border-radius: 0.375rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  --text-opacity: 1;
  color: rgba(107,114,128,var(--text-opacity));
  --bg-opacity: 1;
  background-color: rgba(255,255,255,var(--bg-opacity));
  --border-opacity: 1;
  border-color: rgba(229,231,235,var(--border-opacity));
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.05);
}

.emotion-2:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.emotion-2:hover {
  --bg-opacity: 1;
  background-color: rgba(244,245,247,var(--bg-opacity));
}

<div
    class="emotion-0 emotion-1"
  >
    <button
      class=" emotion-2"
      type="button"
    >
      Button
    </button>
    <button
      class=" emotion-2"
      type="button"
    >
      Button
    </button>
    <button
      class=" emotion-2"
      type="button"
    >
      Button
    </button>
  </div>
</DocumentFragment>
`;

exports[`ButtonGroup attached and not inline 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-0>button:first-of-type {
  border-radius: 0;
  border-top-left-radius: 0.375rem;
  border-bottom-left-radius: 0.375rem;
}

.emotion-0>button:first-of-type:after {
  border-radius: 0;
  border-top-left-radius: 0.5rem;
  border-bottom-left-radius: 0.5rem;
}

.emotion-0>button:last-of-type {
  border-radius: 0;
  border-top-right-radius: 0.375rem;
  border-bottom-right-radius: 0.375rem;
}

.emotion-0>button:last-of-type:after {
  border-radius: 0;
  border-top-right-radius: 0.5rem;
  border-bottom-right-radius: 0.5rem;
}

.emotion-0>button:not(:last-of-type):not(:first-of-type) {
  border-radius: 0;
}

.emotion-0>button:not(:last-of-type):not(:first-of-type):after {
  border-radius: 0;
}

.emotion-0>button:focus {
  z-index: 10;
}

.emotion-0>:not(template)~:not(template) {
  --space-x-reverse: 0;
  margin-right: calc(1rem * var(--space-x-reverse));
  margin-left: calc(1rem * calc(1 - var(--space-x-reverse)));
}

.emotion-0>:not(template)~:not(template) {
  --space-x-reverse: 0;
  margin-right: calc(-1px * var(--space-x-reverse));
  margin-left: calc(-1px * calc(1 - var(--space-x-reverse)));
}

.emotion-2 {
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1.5;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,-webkit-transform;
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  border-width: 1px;
  border-color: transparent;
  border-style: solid;
  cursor: pointer;
  font-family: inherit;
  font-weight: 500;
  border-radius: 0.375rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  --text-opacity: 1;
  color: rgba(107,114,128,var(--text-opacity));
  --bg-opacity: 1;
  background-color: rgba(255,255,255,var(--bg-opacity));
  --border-opacity: 1;
  border-color: rgba(229,231,235,var(--border-opacity));
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.05);
}

.emotion-2:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.emotion-2:hover {
  --bg-opacity: 1;
  background-color: rgba(244,245,247,var(--bg-opacity));
}

<div
    class="emotion-0 emotion-1"
  >
    <button
      class=" emotion-2"
      type="button"
    >
      Button
    </button>
    <button
      class=" emotion-2"
      type="button"
    >
      Button
    </button>
    <button
      class=" emotion-2"
      type="button"
    >
      Button
    </button>
  </div>
</DocumentFragment>
`;

exports[`ButtonGroup not attached and inline 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-0>button:focus {
  z-index: 10;
}

.emotion-0>:not(template)~:not(template) {
  --space-x-reverse: 0;
  margin-right: calc(1rem * var(--space-x-reverse));
  margin-left: calc(1rem * calc(1 - var(--space-x-reverse)));
}

.emotion-2 {
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1.5;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,-webkit-transform;
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  border-width: 1px;
  border-color: transparent;
  border-style: solid;
  cursor: pointer;
  font-family: inherit;
  font-weight: 500;
  border-radius: 0.375rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  --text-opacity: 1;
  color: rgba(107,114,128,var(--text-opacity));
  --bg-opacity: 1;
  background-color: rgba(255,255,255,var(--bg-opacity));
  --border-opacity: 1;
  border-color: rgba(229,231,235,var(--border-opacity));
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.05);
}

.emotion-2:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.emotion-2:hover {
  --bg-opacity: 1;
  background-color: rgba(244,245,247,var(--bg-opacity));
}

<div
    class="emotion-0 emotion-1"
  >
    <button
      class=" emotion-2"
      type="button"
    >
      Button
    </button>
    <button
      class=" emotion-2"
      type="button"
    >
      Button
    </button>
  </div>
</DocumentFragment>
`;

exports[`ButtonGroup not attached and not inline 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0>button:focus {
  z-index: 10;
}

.emotion-0>:not(template)~:not(template) {
  --space-y-reverse: 0;
  margin-top: calc(0.5rem * calc(1 - var(--space-y-reverse)));
  margin-bottom: calc(0.5rem * var(--space-y-reverse));
}

.emotion-2 {
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1.5;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,-webkit-transform;
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  border-width: 1px;
  border-color: transparent;
  border-style: solid;
  cursor: pointer;
  font-family: inherit;
  font-weight: 500;
  border-radius: 0.375rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  --text-opacity: 1;
  color: rgba(107,114,128,var(--text-opacity));
  --bg-opacity: 1;
  background-color: rgba(255,255,255,var(--bg-opacity));
  --border-opacity: 1;
  border-color: rgba(229,231,235,var(--border-opacity));
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.05);
}

.emotion-2:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.emotion-2:hover {
  --bg-opacity: 1;
  background-color: rgba(244,245,247,var(--bg-opacity));
}

<div
    class="emotion-0 emotion-1"
  >
    <button
      class=" emotion-2"
      type="button"
    >
      Button
    </button>
    <button
      class=" emotion-2"
      type="button"
    >
      Button
    </button>
  </div>
</DocumentFragment>
`;
